stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,573 @@
+	c := make(chan error)
+	copyOut := func(w io.Writer, conn net.Conn) {
+		_, err := io.Copy(w, conn)
+		c <- err
+	}
+
+	for _, ep := range endpoints.Targets {
+		d := net.Dialer{Timeout: dialTimeout}
+		conn, err := d.Dial(ep.Domain, ep.Address)
+		if err != nil {
+			return err
+		}
+		defer conn.Close()
+		switch ep.Name {
+		case "stdin":
+			if autoMode || customTargets.stdin {
+				go io.Copy(conn, os.Stdin)
+			}
+		case "stdout":
+			if autoMode || customTargets.stdout {
+				go copyOut(os.Stdout, conn)
+			}
+		case "stderr":
+			if autoMode || customTargets.stderr {
+				go copyOut(os.Stderr, conn)
+			}
+		case "tty":
+			if autoMode || customTargets.ttyIn {
+				go io.Copy(conn, os.Stdin)
+			}
+
+			if autoMode || customTargets.ttyOut {
+				go copyOut(os.Stdout, conn)
+			} else {
+				go copyOut(ioutil.Discard, conn)
+			}
+		}
+	}
+
+	// as soon as one output copying routine fails, this unblocks and the whole process exits
+	err = <-c
+	if err == io.EOF {
+		return nil
+	}
+	return err
+}
+
+// actionPrint prints out available endpoints. This is used by external tools
+// to see which attaching modes are available for an application (eg. `rkt attach --mode=list`)
+func actionPrint(in string, out io.Writer) error {
+	var endpoints Targets
+	statusFile, err := os.OpenFile(in, os.O_RDONLY, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	err = json.NewDecoder(statusFile).Decode(&endpoints)
+	_ = statusFile.Close()
+	if err != nil {
+		return err
+	}
+
+	enc := json.NewEncoder(out)
+	enc.SetIndent("", "    ")
+	return enc.Encode(endpoints)
+}
+
+// actionAttach handles TTY muxing and proxying.
+// It creates a PTY pair and bind-mount the slave to `/rkt/iottymux/<app>/stage2-pts`.
+// Once ready, it sd-notifies as READY so that the main application can be started.
+func actionTTYMux(statusFile string) error {
+	// Open a new TTY pair (master/slave)
+	ptm, pts, err := pty.Open()
+	if err != nil {
+		return err
+	}
+	ttySlavePath := pts.Name()
+	_ = pts.Close()
+	defer ptm.Close()
+	diag.Printf("TTY created, slave pty at %q\n", ttySlavePath)
+
+	// TODO(lucab): set a sane TTY mode here (echo, controls and such).
+
+	// Slave TTY has a dynamic name (eg. /dev/pts/<n>) but a predictable name
+	// is needed, in order to be used as `TTYPath=` value in application unit.
+	// A bind mount is put in place for that, here.
+	ttypath := filepath.Join(pathPrefix, appName, "stage2-pts")
+	f, err := os.Create(ttypath)
+	if err != nil {
+		return err
+	}
+	err = syscall.Mount(pts.Name(), ttypath, "", syscall.MS_BIND, "")
+	if err != nil {
+		return err
+	}
+	// TODO(lucab): double-check this is fine here (alternatives: app-stop or app-rm)
+	defer syscall.Unmount(ttypath, 0)
+	defer f.Close()
+
+	// TODO(lucab): send fd to systemd-manager to ensure we never lose the PTY master fd
+
+	// signal to systemd that the PTY is ready and application can start
+	ok, err := daemon.SdNotify(true, "READY=1")
```

Comment:
This should be in line with SdNotify signature: https://github.com/coreos/go-systemd/blob/v14/daemon/sdnotify.go#L28-L31

In this case we require sd-notification to proceed, so message non-delivery is an hard failure here.